### PR TITLE
[fix] Auth Override not working with DefinitionBody fix

### PR DESCRIPTION
**Pull request by paulhcsun**
### Description of changes
Adding new optional opt-in property `OverrideApiAuth` with a `boolean` value to the `AWS::Serverless::Function` `Events` level. 

### Description of how you validated changes
Transformation test

### Checklist

- [x] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [x] Add/update [transform tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions)
    - [x] Using correct values
    - [ ] Using wrong values
- [ ] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)

### Examples?

Please reach out in the comments if you want to add an example. Examples will be 
added to `sam init` through [aws/aws-sam-cli-app-templates](https://github.com/aws/aws-sam-cli-app-templates).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samtranslator/internal/schema_source/aws_serverless_function.py
Comment (aaythapa) [inline samtranslator/internal/schema_source/aws_serverless_function.py:284]: Why's it here aswell? Isn't it just under auth?

File: samtranslator/model/eventsources/push.py
Comment (aaythapa) [inline samtranslator/model/eventsources/push.py:672]: Same comment as above, Shouldn't this go under the list of [auth properties](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-property-function-apifunctionauth.html)?

File: samtranslator/model/sam_resources.py
Comment (aaythapa) [inline samtranslator/model/sam_resources.py:182]: Don't need to add this here, this is only for list of SAM function properites

File: tests/model/eventsources/test_api_event_source.py
Comment (aaythapa) [inline tests/model/eventsources/test_api_event_source.py:29]: what's the reason for messing with kwargs here (can't deduce from the code) rather than testing it how it was before?

File: tests/translator/input/api_with_swagger_authorizer_none_test_auth_override.yaml
Comment (aaythapa) [inline tests/translator/input/api_with_swagger_authorizer_none_test_auth_override.yaml:112]: having the events name state which ones overriding vs which ones not would help to easily compare with the output

File: tests/translator/input/api_with_swagger_authorizer_none_test_auth_override.yaml
Comment (aaythapa) [inline tests/translator/input/api_with_swagger_authorizer_none_test_auth_override.yaml:1]: this test looks a little bloated, lot of the template isn't needed or relevant to what you're testing

File: tests/model/eventsources/test_api_event_source.py
Comment (paulhcsun) [reply tests/model/eventsources/test_api_event_source.py:29]: This is because we changed the way we get the `intrinsics_resolver` in the `to_cloudformation()` method from `intrinsics_resolver = kwargs.get("intrinsics_resolver")` to `intrinsics_resolver: IntrinsicsResolver = kwargs["intrinsics_resolver"]` and caused those tests to fail because we now expect there to always be an `IntrinsicsResolver`. We changed this because the `self.add_auth_to_swagger()` call compalins about `intrinsics_resolver` being returned as `Optional` by the `.get()` method. It is returned similarly in the `sam_resources.py` file.

File: samtranslator/model/sam_resources.py
Comment (paulhcsun) [reply samtranslator/model/sam_resources.py:182]: Good catch, i'll remove this.

File: samtranslator/internal/schema_source/aws_serverless_function.py
Comment (paulhcsun) [reply samtranslator/internal/schema_source/aws_serverless_function.py:284]: Good catch, will remove this.

File: samtranslator/model/eventsources/push.py
Comment (GavinZZ) [inline samtranslator/model/eventsources/push.py:750]: I thought the name `TestAuthOverride` is still pending UX decision based on the doc I saw yesterday. Has that decision been made?

File: tests/translator/input/api_with_swagger_authorizer_none_test_auth_override.yaml
Comment (paulhcsun) [reply tests/translator/input/api_with_swagger_authorizer_none_test_auth_override.yaml:130]: trimmed these down so this should be fine now as there is just one event per api.

File: samtranslator/model/eventsources/push.py
Comment (paulhcsun) [reply samtranslator/model/eventsources/push.py:748]: Not yet, I've signed up for tomorrows UX office hour and will push a change to update the property name once that decision is made. The current PR is open just for general review and will not be merged until the property name is updated.

File: samtranslator/model/eventsources/push.py
Comment (bentvelj) [inline samtranslator/model/eventsources/push.py:748]: [nit] Is the `is True` necessary?

Comment (paulhcsun) [reply samtranslator/model/eventsources/push.py:748]: Good catch, I don't think it is. I'll push a change to remove this.

File: samtranslator/model/eventsources/push.py
Comment (ssenchenko) [inline samtranslator/model/eventsources/push.py:730]: Why haven't we been sure that it's not `None` but now we are?

File: tests/model/eventsources/test_api_event_source.py
Comment (ssenchenko) [reply tests/model/eventsources/test_api_event_source.py:29]: So we're doing that only because of the static type check?
Then we should initialize `intrinsic_resolver` to an empty object in `push.py` and don't touch the tests as it can be a backward compat issue

**Review by GavinZZ — APPROVED** [review]
Changes LGTM, but my understanding is that we won't merge this until UX review. Just want to make sure we don't accidentally merge it.

File: samtranslator/model/eventsources/push.py
Comment (ssenchenko) [reply samtranslator/model/eventsources/push.py:730]: As I see below, we probably have to create an empty `IntrinsicResolver` here if we need it but... still we need to be very cautious about that. Is it `self.add_auth_to_swagger` which requires the resolver not to be `None`? Can we handle `None` value inside it?

File: samtranslator/model/eventsources/push.py
Comment (aaythapa) [inline samtranslator/model/eventsources/push.py:763]: Can you add some comments here? Bit confused about what its actually doing

File: tests/model/eventsources/test_api_event_source.py
Comment (aaythapa) [reply tests/model/eventsources/test_api_event_source.py:29]: > We changed this because the self.add_auth_to_swagger() call compalins about intrinsics_resolver being returned as Optional by the .get() method.

what was complaining exactly? (if you have the err msg that'd be helpful)

File: tests/translator/input/api_with_swagger_authorizer_none_override_api_auth.yaml
Comment (aaythapa) [inline tests/translator/input/api_with_swagger_authorizer_none_override_api_auth.yaml:1]: tests look good, can you add one more event without `OverrideApiAuth` just to show what output looks like without it being overridden

Comment (paulhcsun) [reply tests/translator/input/api_with_swagger_authorizer_none_override_api_auth.yaml:1]: Good idea, i'll add that in. Also in response to the above comment, here is the error for `intrinsics_resolver`:

`samtranslator/model/eventsources/push.py:762: error: Argument 9 to "add_auth_to_swagger" of "Api" has incompatible type "Optional[Any]"; expected "IntrinsicsResolver"  [arg-type]`

File: samtranslator/model/eventsources/push.py
Comment (paulhcsun) [reply samtranslator/model/eventsources/push.py:763]: For sure, i'll add some comments in the next commit. Essentially it's just creating the resources needed to call `self.add_auth_to_swagger()` and checking if the `OverrideApiAuth` property is set. 

`add_auth_to_swagger()` is needed because it will call `add_auth_to_method()` in `swagger.py` which checks event source `Auth` and updates the `DefinitionBody` with any security overrides.

then the line `explicit_api["DefinitionBody"] = editor.swagger` saves that change to `DefinitionBody` so it can be detected when transforming the `Api` resource.

**Comment by paulhcsun** [comment]
> Changes LGTM, but my understanding is that we won't merge this until UX review. Just want to make sure we don't accidentally merge it.

UX review is finished, the property name has been approved and is okay to merge once this PR is approved. I will remove the [DO NOT MERGE].

File: samtranslator/model/eventsources/push.py
Comment (paulhcsun) [reply samtranslator/model/eventsources/push.py:730]: `self.add_auth_to_swagger` expects `IntrinsicsResolver` but the `.get()` is returning an `Optional[Any]`. 

I don't think we can handle this inside of `add_auth_to_swagger()` but it seems like if i add a check on `intrinsics_resolver` before calling `add_auth_to_swagger()` it is happy.

i.e. `if swagger_body and self.Auth and intrinsics_resolver and self.Auth.get("OverrideApiAuth"):`

Comment (ssenchenko) [reply samtranslator/model/eventsources/push.py:730]: Sounds good. But do we really need an intrinsic resolver to be non null to make override happen?

**Comment by paulhcsun** [comment]
Due to [limited support for intrinsic functions in SAM templates](https://github.com/aws/serverless-application-model/issues/2533), we will not be supporting intrinsics for the new property `OverrideApiAuth`.

**Review by aaythapa — APPROVED** [review]

**Review by ssenchenko — APPROVED** [review]

**Pull request merged**
